tool/smarts/src/main/java/org/openscience/cdk/smarts/SmartsFragmentExtractor.java

```diff
@@ -0,0 +1,455 @@
+/*
+ * Copyright (c) 2016 John May <jwmay@users.sf.net>
+ *
+ * Contact: cdk-devel@lists.sourceforge.net
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or (at
+ * your option) any later version. All we ask is that proper credit is given
+ * for our work, which includes - but is not limited to - adding the above
+ * copyright notice to the beginning of your source code files, and to any
+ * copyright notice that you may distribute with programs based on this work.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 U
+ */
+
+package org.openscience.cdk.smarts;
+
+import org.openscience.cdk.interfaces.IAtom;
+import org.openscience.cdk.interfaces.IAtomContainer;
+import org.openscience.cdk.interfaces.IBond;
+
+import java.util.Arrays;
+import java.util.Locale;
+
+/**
+ * Utility class to create SMARTS that match part (substructure) of a molecule.
+ * SMARTS are generated by providing the atom indexes. An example use cases is
+ * encoding features from a fingerprint.
+ * <p>
+ * The extractor has two modes. {@link #MODE_EXACT} (default) captures the element,
+ * valence, hydrogen count, connectivity, and charge in the SMARTS atom expressions.
+ * The alternative mode, {@link #MODE_JCOMPOUNDMAPPER}, only captures the element,
+ * non-zero charge, and peripheral bonds. Although the later looks cleaner, the
+ * peripheral bonds intend to capture the connectivity of the terminal atoms but
+ * since the valence is not bounded further substitution is still allowed.
+ *
+ * <p>The difference is easily demonstrated for methyl. Consider the compound
+ * of 2-methylpentane {@code CC(C)CCC}, if we extract one of the methyl atoms
+ * depending on the mode we obtain {@code [CH3v4X4+0]} or {@code C*}. The first
+ * of these patterns (obtained by {@link #MODE_EXACT}) matches the compound in
+ * <b>three places</b> (the three methyl groups). The second matches <b>six</b>
+ * times (every atom) because the substituion on the carbon is not locked.
+ * A further complication is introduced by the inclusion of the peripheral atoms,
+ * for 1H-indole {@code [nH]1ccc2c1cccc2} we can obtain the SMARTS {@code n(ccc(a)a)a}
+ * that doesn't match at all. This is because one of the aromatic atoms ('a')
+ * needs to match the nitrogen.
+ *
+ * <p><b>Basic Usage:</b></p>
+ * <pre>{@code
```

Comment:
I think that's the only difference, you can find them with this:

```
find ~/workspace/github/cdk -name '*.java' -exec grep -H '&lt;' {} \;
```

```
/Users/john/workspace/github/cdk/cdk/app/depict/src/test/java/org/openscience/cdk/depict/SvgDrawVisitorTest.java:                                          + "    <text  x='50.0' y='50.0' fill='#FF0000' text-anchor='middle'>PNG &lt; EPS &lt; SVG</text>\n"
org/openscience/cdk/graph/AllPairsShortestPaths.java: * for (int i = 0; i &lt; benzene.getAtomCount(); i++) {
org/openscience/cdk/graph/AllPairsShortestPaths.java: *     for (int j = i + 1; j &lt; benzene.getAtomCount(); j++) {
org/openscience/cdk/graph/PathTools.java:     * If number of atoms in or below sphere x&lt;max and number of atoms in or below sphere x+1&gt;max then
org/openscience/cdk/graph/TripletShortCycles.java: * ESSSR <br/>(fingerprints only store cycles |C| &lt;=
org/openscience/cdk/isomorphism/Mappings.java:     *     Iterable&lt;String&gt; strs = mappings.map(new Function&lt;int[], String&gt;() {
org/openscience/cdk/isomorphism/Mappings.java:     *             for (int i = 0; i &lt; input.length; i++) {
org/openscience/cdk/isomorphism/Mappings.java:     * for (Map&lt;IAtom,IAtom&gt; map : mappings.toAtomMap()) {
org/openscience/cdk/isomorphism/Mappings.java:     *     for (Map.Entry&lt;IAtom,IAtom&gt; e : map.entrySet()) {
org/openscience/cdk/isomorphism/Mappings.java:     * for (Map&lt;IBond,IBond&gt; map : mappings.toBondMap()) {
org/openscience/cdk/isomorphism/Mappings.java:     *     for (Map.Entry&lt;IBond,IBond&gt; e : map.entrySet()) {
org/openscience/cdk/isomorphism/Mappings.java:     * for (Map&lt;IChemObject,IChemObject&gt; map : mappings.toBondMap()) {
org/openscience/cdk/isomorphism/Mappings.java:     *     for (Map.Entry&lt;IChemObject,IChemObject&gt; e : map.entrySet()) {
org/openscience/cdk/reaction/ReactionSpecification.java:     *          expected to be &lt;dictionaryNameSpace&gt;:&lt;entryID&gt;.
org/openscience/cdk/qsar/DescriptorSpecification.java:     *          expected to be &lt;dictionaryNameSpace&gt;:&lt;entryID&gt;.
org/openscience/cdk/qsar/DescriptorSpecification.java:     *          expected to be &lt;dictionaryNameSpace&gt;:&lt;entryID&gt;.
org/openscience/cdk/signature/MoleculeSignature.java: * List&lt;Orbit&gt; orbits = moleculeSignature.calculateOrbits();
/Users/john/workspace/github/cdk/cdk/doc/javadoc/source/XMIDoclet.java:     * into &lt;UML:Class> elements.
/Users/john/workspace/github/cdk/cdk/doc/javadoc/source/XMIDoclet.java:     * Method that serializes ClassDoc objects into &lt;UML:Class> elements.
/Users/john/workspace/github/cdk/cdk/doc/javadoc/source/XMIDoclet.java:     * into &lt;listitem> elements (Umbrello specific)?.
/Users/john/workspace/github/cdk/cdk/doc/javadoc/source/XMIDoclet.java:     * Method that serializes ClassDoc objects into &lt;listitem> elements
org/openscience/cdk/math/qm/GaussiansBasis.java: * S = &lt;phi_i|phi_j><br>
org/openscience/cdk/math/qm/GaussiansBasis.java: * J = &lt;d/dr phi_i | d/dr phi_j><br>
org/openscience/cdk/math/qm/GaussiansBasis.java: * V = &lt;phi_i | 1/r | phi_j><br>
org/openscience/cdk/math/qm/IBasis.java:     * Calculate the overlap integral S = &lt;phi_i|phi_j>.
org/openscience/cdk/math/qm/IBasis.java:     * Calculates the impulse J = -&lt;d/dr chi_i | d/dr chi_j>.
org/openscience/cdk/math/qm/IBasis.java:     * Calculates the potential V = &lt;chi_i | 1/r | chi_j>.
org/openscience/cdk/geometry/alignment/KabschAlignment.java: * for (int i = 0; i &lt; ac1.getAtomCount(); i++) {
org/openscience/cdk/io/MDLRXNWriter.java:     * &gt; &lt;key&gt;<br>
org/openscience/cdk/io/MDLV2000Reader.java:     * not currently return field numbers (e.g. DT&lt;n&gt;).
org/openscience/cdk/io/ReaderFactory.java: *   StringReader stringReader = "&lt;molecule/>";
/Users/john/workspace/github/cdk/cdk/storage/io/src/test/java/org/openscience/cdk/io/cml/JmolTest.java:     * <ul><li> &lt;crystal></li></ul>
org/openscience/cdk/io/FormatFactory.java: *   StringReader stringReader = new StringReader("&lt;molecule/>");
org/openscience/cdk/libio/cml/Convertor.java:     * @param useCMLIDs Uses object IDs like 'a1' instead of 'a&lt;hash>'.
org/openscience/cdk/normalize/Normalizer.java:     *  &lt;replace-set&gt;<br>
org/openscience/cdk/normalize/Normalizer.java:     *  &lt;replace&gt;O=N=O&lt;/replace&gt;<br>
org/openscience/cdk/normalize/Normalizer.java:     *  &lt;replacement&gt;[O-][N+]=O&lt;/replacement&gt;<br>
org/openscience/cdk/normalize/Normalizer.java:     *  &lt;/replace-set&gt;<br>
org/openscience/cdk/tools/manipulator/MolecularFormulaManipulator.java:     * System with numbers wrapped in &lt;sub&gt;&lt;/sub&gt; tags. Useful for
org/openscience/cdk/tools/manipulator/MolecularFormulaManipulator.java:     * System with numbers wrapped in &lt;sub&gt;&lt;/sub&gt; tags and the
org/openscience/cdk/tools/manipulator/MolecularFormulaManipulator.java:     * isotope of each Element in &lt;sup&gt;&lt;/sup&gt; tags and the total
org/openscience/cdk/tools/manipulator/MolecularFormulaManipulator.java:     * charge of IMolecularFormula in &lt;sup&gt;&lt;/sup&gt; tags. Useful for
org/openscience/cdk/tools/manipulator/MolecularFormulaManipulator.java:     * wrapped in &lt;sub&gt;&lt;/sub&gt; tags and the isotope of each Element
org/openscience/cdk/tools/manipulator/MolecularFormulaManipulator.java:     * in &lt;sup&gt;&lt;/sup&gt; tags and the total showCharge of IMolecularFormula
org/openscience/cdk/tools/manipulator/MolecularFormulaManipulator.java:     * in &lt;sup&gt;&lt;/sup&gt; tags. Useful for displaying formulae in Swing
org/openscience/cdk/smiles/smarts/SMARTSQueryTool.java: *    for (int i = 0; i &lt; nmatch; i++) {
org/openscience/cdk/smiles/smarts/SMARTSQueryTool.java: * by Craig James the <code>h&lt;n&gt;</code> SMARTS pattern should not be used. It was included in the Daylight spec
org/openscience/cdk/smiles/smarts/SMARTSQueryTool.java: * for backwards compatibility. To match hydrogens, use the <code>H&lt;n&gt;</cod> pattern.</li> <li>The wild card
net/sf/cdk/tools/bibtex/BibTeXMLFile.java:	 * Returns an Iterator&lt;BibTeXMLEntry>.
```